lua/orgmode/objects/date.lua

```diff
@@ -517,17 +518,45 @@ end
 ---@return OrgDate
 function Date:add(opts)
   opts = opts or {}
+  ---@type table
   local date = os.date('*t', self.timestamp)
   for opt, val in pairs(opts) do
     if opt == 'week' then
       opt = 'day'
       val = val * 7
     end
     date[opt] = date[opt] + val
+    if opt == 'month' then
+      local days_of_month = Date._days_of_month(date)
+      if days_of_month < date['day'] then
+        date['day'] = days_of_month
+      end
+    end
   end
   return self:from_time_table(date)
 end
 
+---@param date table
+---@return number
+function Date._days_of_month(date)
+  local long_months = { true, false, true, false, true, false, true, true, false, true, false, true }
+  if date.month == 2 then
+    return Date._days_of_february(date.year)
+  elseif long_months[date.month] then
+    return 31
+  else
+    return 30
+  end
+end
+
+function Date._days_of_february(year)
+  return Date._is_leap_year(year) and 29 or 28
+end
+
+function Date._is_leap_year(year)
+  return year % 400 == 0 or not (year % 100 == 0) and year % 4 == 0
```

Comment:
Let's add some parenthesis around these so it's easier to read:
 
```suggestion
  return year % 400 == 0 or (year % 100 ~= 0 and year % 4 == 0)
```

Reply:
Done.